File: backend-python/app/providers/aws/search_store.py
from __future__ import annotations

import json
import os
import urllib.request
from typing import Any, Dict, List
from urllib.parse import quote

try:
    from opensearchpy import OpenSearch, RequestsHttpConnection
    from requests_aws4auth import AWS4Auth
    import boto3
except ImportError:  # pragma: no cover
    OpenSearch = None  # type: ignore
    AWS4Auth = None  # type: ignore
    boto3 = None  # type: ignore

from ..base import SearchStoreProvider


def _opensearch_config() -> tuple[str, str]:
    endpoint = os.getenv("OPENSEARCH_ENDPOINT", "").strip().rstrip("/")
    index_name = os.getenv("OPENSEARCH_INDEX_NAME", "rag-chunks").strip()
    return endpoint, index_name


def _opensearch_client():
    if OpenSearch is None:
        raise RuntimeError("opensearch-py and requests-aws4auth are required for AWS search.")
    endpoint, _ = _opensearch_config()
    region = os.getenv("AWS_REGION", "ap-southeast-2")

    if boto3 is not None:
        credentials = boto3.Session().get_credentials()
        awsauth = AWS4Auth(
            credentials.access_key,
            credentials.secret_key,
            region,
            "es",
            session_token=credentials.token,
        )
        return OpenSearch(
            hosts=[{"host": endpoint.replace("https://", "").replace("http://", ""), "port": 443}],
            http_auth=awsauth,
            use_ssl=True,
            verify_certs=True,
            connection_class=RequestsHttpConnection,
        )
    raise RuntimeError("boto3 is required for AWS OpenSearch auth.")


class AwsSearchStoreProvider(SearchStoreProvider):
    def is_configured(self) -> bool:
        endpoint, _ = _opensearch_config()
        return bool(endpoint)

    def upsert_chunks(
        self,
        tenant_id: str,
        document_id: str,
        blob_name: str,
        file_name: str,
        chunks: List[str],
    ) -> None:
        if not self.is_configured() or not chunks:
            return
        try:
            client = _opensearch_client()
            _, index_name = _opensearch_config()
            bulk_body = []
            for idx, content in enumerate(chunks):
                doc_id = f"{document_id}-{idx}"
                bulk_body.append({"index": {"_index": index_name, "_id": doc_id}})
                bulk_body.append({
                    "id": doc_id,
                    "tenantId": tenant_id,
                    "documentId": document_id,
                    "blobName": blob_name,
                    "fileName": file_name,
                    "chunkIndex": idx,
                    "content": content,
                    "contentLength": len(content),
                    "sourceType": "python-backend",
                })
            client.bulk(body=bulk_body)
        except Exception:
            pass  # best-effort

    def delete_chunks_for_document(self, document_id: str, tenant_id: str) -> int:
        if not self.is_configured():
            return 0
        try:
            client = _opensearch_client()
            _, index_name = _opensearch_config()
            response = client.delete_by_query(
                index=index_name,
                body={
                    "query": {
                        "bool": {
                            "must": [
                                {"term": {"documentId.keyword": document_id}},
                                {"term": {"tenantId.keyword": tenant_id}},
                            ]
                        }
                    }
                },
            )
            return response.get("deleted", 0)
        except Exception:
            return 0

    def list_document_groups(self, tenant_id: str) -> List[Dict[str, Any]]:
        if not self.is_configured():
            return []
        try:
            client = _opensearch_client()
            _, index_name = _opensearch_config()
            response = client.search(
                index=index_name,
                body={
                    "size": 0,
                    "query": {"term": {"tenantId.keyword": tenant_id}},
                    "aggs": {
                        "docs": {
                            "terms": {"field": "documentId.keyword", "size": 300},
                            "aggs": {
                                "fileName": {"terms": {"field": "fileName.keyword", "size": 1}},
                                "blobName": {"terms": {"field": "blobName.keyword", "size": 1}},
                                "chunkCount": {"value_count": {"field": "chunkIndex"}},
                            },
                        }
                    },
                },
            )
            results = []
            for bucket in response.get("aggregations", {}).get("docs", {}).get("buckets", []):
                file_names = bucket.get("fileName", {}).get("buckets", [])
                blob_names = bucket.get("blobName", {}).get("buckets", [])
                results.append({
                    "documentId": bucket["key"],
                    "fileName": file_names[0]["key"] if file_names else "",
                    "blobName": blob_names[0]["key"] if blob_names else "",
                    "chunkCount": bucket.get("chunkCount", {}).get("value", 0),
                })
            return results
        except Exception:
            return []

File: backend-nodejs/functions-ingestion/src/providers/aws/searchStore.ts
import { Client as OpenSearchClient } from "@opensearch-project/opensearch";
import { AwsSigv4Signer } from "@opensearch-project/opensearch/aws";
import { defaultProvider } from "@aws-sdk/credential-provider-node";
import type { SearchStoreProvider } from "../base.js";

function isIndexNotFound(error: unknown): boolean {
  if (!(error instanceof Error)) {
    return false;
  }

  return error.message.includes("index_not_found_exception");
}

function getClient(): OpenSearchClient {
  const endpoint = process.env.OPENSEARCH_ENDPOINT;
  if (!endpoint) throw new Error("OPENSEARCH_ENDPOINT is required");

  const normalizedEndpoint = endpoint.toLowerCase();
  const isManagedOpenSearch =
    normalizedEndpoint.includes(".es.amazonaws.com") ||
    normalizedEndpoint.includes(".aoss.amazonaws.com");

  if (!isManagedOpenSearch) {
    return new OpenSearchClient({
      node: endpoint
    });
  }

  const region = process.env.AWS_REGION ?? "ap-southeast-2";
  return new OpenSearchClient({
    ...AwsSigv4Signer({
      region,
      service: "es",
      getCredentials: defaultProvider()
    }),
    node: endpoint
  });
}

let _client: OpenSearchClient | undefined;
function client(): OpenSearchClient {
  if (!_client) _client = getClient();
  return _client;
}

const INDEX = () => process.env.OPENSEARCH_INDEX_NAME ?? "rag-chunks";

export class AwsSearchStoreProvider implements SearchStoreProvider {
  isEnabled(): boolean {
    // If OPENSEARCH_ENDPOINT is present, default SEARCH_ENABLED to true for AWS.
    // Otherwise, respect explicit SEARCH_ENABLED or default to false.
    const endpoint = process.env.OPENSEARCH_ENDPOINT?.trim();
    if (!endpoint) {
      return false;
    }

    const hasEndpoint = Boolean(endpoint);
    const searchEnabledEnv = process.env.SEARCH_ENABLED?.trim().toLowerCase();

    // If endpoint exists and SEARCH_ENABLED is not explicitly false, enable search.
    if (searchEnabledEnv === "false") {
      return false;
    }

    return hasEndpoint;
  }

  async indexChunks(chunks: Record<string, unknown>[]): Promise<boolean> {
    if (!this.isEnabled()) return false;
    if (!chunks.length) return false;
    const body = chunks.flatMap(doc => [
      { index: { _index: INDEX(), _id: String(doc.id) } },
      doc
    ]);
    await client().bulk({ body });
    return true;
  }

  async searchChunks(
    query: string,
    tenantId: string,
    top = 3,
    queryEmbedding?: number[],
    mode = "hybrid"
  ): Promise<Record<string, unknown>[]> {
    if (!this.isEnabled()) return [];
    const mustFilter = { term: { "tenantId.keyword": tenantId } };

    let body: Record<string, unknown>;
    if (mode === "vector" && queryEmbedding) {
      body = {
        query: { bool: { filter: [mustFilter] } },
        knn: { embedding: { vector: queryEmbedding, k: top } },
        size: top
      };
    } else if (mode === "hybrid" && queryEmbedding) {
      body = {
        query: {
          bool: {
            must: [{ match: { content: query } }],
            filter: [mustFilter]
          }
        },
        knn: { embedding: { vector: queryEmbedding, k: top } },
        size: top
      };
    } else {
      body = {
        query: {
          bool: {
            must: [{ match: { content: query } }],
            filter: [mustFilter]
          }
        },
        size: top
      };
    }

    try {
      const response = await client().search({ index: INDEX(), body });
      return (response.body.hits?.hits ?? []).map(
        (h: Record<string, unknown>) => ({
          ...(h._source as Record<string, unknown>),
          score: h._score
        })
      );
    } catch (error) {
      if (isIndexNotFound(error)) {
        return [];
      }

      throw error;
    }
  }

  async deleteChunksForDocument(
    documentId: string,
    tenantId: string
  ): Promise<number> {
    if (!this.isEnabled()) return 0;
    try {
      const response = await client().deleteByQuery({
        index: INDEX(),
        body: {
          query: {
            bool: {
              filter: [
                { term: { "documentId.keyword": documentId } },
                { term: { "tenantId.keyword": tenantId } }
              ]
            }
          }
        }
      });
      return (response.body.deleted as number) ?? 0;
    } catch (error) {
      if (isIndexNotFound(error)) {
        return 0;
      }

      throw error;
    }
  }

  async countChunksForDocument(
    documentId: string,
    tenantId: string
  ): Promise<number> {
    if (!this.isEnabled()) return 0;
    try {
      const response = await client().count({
        index: INDEX(),
        body: {
          query: {
            bool: {
              filter: [
                { term: { "documentId.keyword": documentId } },
                { term: { "tenantId.keyword": tenantId } }
              ]
            }
          }
        }
      });
      return response.body.count ?? 0;
    } catch (error) {
      if (isIndexNotFound(error)) {
        return 0;
      }

      throw error;
    }
  }

  async listDocumentGroups(
    tenantId: string
  ): Promise<Record<string, unknown>[]> {
    if (!this.isEnabled()) return [];
    try {
      const response = await client().search({
        index: INDEX(),
        body: {
          query: { term: { "tenantId.keyword": tenantId } },
          size: 0,
          aggs: {
            docs: {
              terms: { field: "documentId.keyword", size: 500 },
              aggs: {
                fileName: { terms: { field: "fileName.keyword", size: 1 } },
                blobName: { terms: { field: "blobName.keyword", size: 1 } }
              }
            }
          }
        }
      });
      return (response.body.aggregations?.docs?.buckets ?? []).map(
        (b: Record<string, unknown>) => ({
          documentId: b.key,
          chunkCount: b.doc_count,
          fileName:
            (
              (b.fileName as Record<string, unknown[]>)?.buckets?.[0] as Record<
                string,
                unknown
              >
            )?.key ?? "",
          blobName:
            (
              (b.blobName as Record<string, unknown[]>)?.buckets?.[0] as Record<
                string,
                unknown
              >
            )?.key ?? ""
        })
      );
    } catch (error) {
      if (isIndexNotFound(error)) {
        return [];
      }

      throw error;
    }
  }
}
